Web Inspector: Add editing capabilities for font variation axes in Fonts sidebar panel

rcaliman-apple opened the pull request:
#### b948d0181bcf2490758c678c97fac81e9392fbbe
<pre>
Web Inspector: Add editing capabilities for font variation axes in Fonts sidebar panel
<a href="https://bugs.webkit.org/show_bug.cgi?id=249535">https://bugs.webkit.org/show_bug.cgi?id=249535</a>

Reviewed by Patrick Angle.

Variable fonts expose variation axes to control specific aspects of the font,
like thickness, stretch or slant. The Fonts sidebar panel already lists the
available variation axes for the primary font used on the selected node.

This patch augments the Fonts sidebar panel with editing controls to interactively
adjust the values for variation axes and get immediate visual feedback of the results.

There are two types of variation axes: registered and custom.

Registered axes values map to a corresponding CSS font property:
`wght` -&gt; font-weight
`wdth` -&gt; font-stretch
`ital` -&gt; font-style
`slnt` -&gt; font-style

Custom axes and their values are always written to `font-variation-settings`.

If a registered axis is declared in the value of `font-variation-settings`,
its value must always be written there instead of the corresponding font property
else it will be overriden by the value of `font-variation-settings`.

In this implementation, all edits made using the controls in the Fonts sidebar panel
are written to the selected node&apos;s inline style to ensure immediate visual feedback
at the expense of scoping the font style changes very narrowly to the selected node only.

A future iteration will improve this by identifying the appropriate CSS rules
in the cascade to write to.

* LayoutTests/inspector/model/font-calculate-properties-expected.txt:
* LayoutTests/inspector/model/font-calculate-properties.html:
* LayoutTests/inspector/model/font-styles-write-variation-expected.txt: Added.
* LayoutTests/inspector/model/font-styles-write-variation.html: Added.

Updated existing test to use the new `WI.FontStyles` model.
Added new test to validate variation axis writing logic.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Models/Font.js:
(WI.Font.prototype.variationAxis):
(WI.Font):
(WI.Font.prototype.calculateFontProperties): Deleted.
(WI.Font.prototype._calculateProperties): Deleted.
(WI.Font.prototype._calculateFontFeatureAxes): Deleted.
(WI.Font.prototype._calculateFontVariationAxes): Deleted.
(WI.Font.prototype._parseFontFeatureOrVariationSettings): Deleted.

Moved parsing-related logic from `WI.Font` to a new `WI.FontStyles` model.
This object maintains state of edits made to variation axes over time.
It handles where to write variation axis values.

* Source/WebInspectorUI/UserInterface/Models/FontStyles.js: Copied from Source/WebInspectorUI/UserInterface/Models/Font.js.
(WI.FontStyles):
(WI.FontStyles.prototype.get featuresMap):
(WI.FontStyles.prototype.get variationsMap):
(WI.FontStyles.prototype.get propertiesMap):
(WI.FontStyles.prototype.get significantChangeSinceLastRefresh):
(WI.FontStyles.fontPropertyForAxisTag):
(WI.FontStyles.axisValueToFontPropertyValue):
(WI.FontStyles.fontPropertyValueToAxisValue):
(WI.FontStyles.prototype.writeFontVariation):
(WI.FontStyles.prototype.refresh):
(WI.FontStyles.prototype._calculateFontProperties):
(WI.FontStyles.prototype._calculateProperties):
(WI.FontStyles.prototype._calculateFontFeatureAxes):
(WI.FontStyles.prototype._calculateFontVariationAxes):
(WI.FontStyles.prototype._parseFontFeatureOrVariationSettings):

Most contents of `WI.FontStyles` are copied over from `WI.Font` verbatim.

* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section &gt; .content &gt; .group:has(.row.font-variation)):
(.details-section &gt; .content &gt; .group &gt; .row:is(.simple, .font-variation):has(.warning)):
(.details-section &gt; .content &gt; .group &gt; .row:is(.simple, .font-variation) &gt; .warning):
(.details-section &gt; .content &gt; .group &gt; .row.simple:has(.warning)): Deleted.
(.details-section &gt; .content &gt; .group &gt; .row.simple &gt; .warning): Deleted.
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel):
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.detached):
(WI.FontDetailsPanel.prototype.layout):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype._createDetailsSectionRowForProperty):
(WI.FontDetailsPanel.prototype._formatPropertyValue):
(WI.FontDetailsPanel.prototype._handleFontVariationValueChanged):
(WI.FontDetailsPanel.prototype._formatSizeValue): Deleted.
(WI.FontDetailsPanel.prototype._formatStyleValue): Deleted.
(WI.FontDetailsPanel.prototype._formatSimpleSingleValue): Deleted.
(WI.FontDetailsPanel.prototype._formatVariationValue): Deleted.
(WI.FontDetailsPanel.prototype._createVariationValueElement): Deleted.
(WI.FontDetailsPanel.prototype._hasVariationValue): Deleted.

Inside `FontDetailsPanel.layout()` the structure for editing controls is created.
Inside `FontDetailsPanel.refresh()` the values for editing controls are updated.
This ensures the UI remains in sync when changes occur in the Styles sidebar panel
or new styles apply as a result of DOM mutations, matching media queries, etc.

* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css: Added.
(.details-section &gt; .content &gt; .group &gt; .row.font-variation):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .label:not(:empty)):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .tag):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .value):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]::-webkit-slider-thumb):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]::-webkit-slider-thumb:active):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]::-webkit-slider-runnable-track):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]::before,):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; input[type=&quot;range&quot;]::after):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; .variation-minvalue):
(.details-section &gt; .content &gt; .group &gt; .row.font-variation &gt; .variation-range &gt; .variation-maxvalue):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js: Added.
(WI.FontVariationDetailsSectionRow):
(WI.FontVariationDetailsSectionRow.prototype.get tag):
(WI.FontVariationDetailsSectionRow.prototype.set tag):
(WI.FontVariationDetailsSectionRow.prototype.get label):
(WI.FontVariationDetailsSectionRow.prototype.set label):
(WI.FontVariationDetailsSectionRow.prototype.get value):
(WI.FontVariationDetailsSectionRow.prototype.set value):
(WI.FontVariationDetailsSectionRow.prototype.get warningMessage):
(WI.FontVariationDetailsSectionRow.prototype.set warningMessage):
(WI.FontVariationDetailsSectionRow.prototype._formatAxisValueAsString):
(WI.FontVariationDetailsSectionRow.prototype._getAxisResolution):

`WI.FontVariationDetailsSectionRow` encapsulates the layout for editing controls:
the interactive slider, the track with min/max values, warnings for out-of-range values.

* Source/WebInspectorUI/UserInterface/Views/Variables.css:
(:root):
(@media (prefers-color-scheme: dark) :root):

Canonical link: <a href="https://commits.webkit.org/258503@main">https://commits.webkit.org/258503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/337aa7d405c2df1dd534983e8baf7ea63abe3155

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102147 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35218 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111472 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12271 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2205 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109207 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107926 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9396 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92669 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24150 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4843 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25574 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4959 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11011 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45071 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5833 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6704 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on LayoutTests/inspector/model/font-styles-write-variation.html:
```diff
@@ -0,0 +1,180 @@
+            name,
+            description,
+            async setup() {
+                return loadFontFace("normal normal 12px VariableFont", " ", "TestPage-VariableFont1Loaded");
```
I suspect the other test is using "return" because I had originally used promises instead of async functions in an earlier version of the patch that added the test, but it might make more sense for this to be `await loadFontFace(...);` since we are in an async context?

patrickangle commented on LayoutTests/inspector/model/font-styles-write-variation.html:
```diff
@@ -0,0 +1,180 @@
+        });
+    }
+
+    function loadFontFace(fontDeclaration, text, eventName)
```
Nit: We should declare this before `function addTestCase(...` since it is used by `addTestCase`.

patrickangle commented on LayoutTests/inspector/model/font-styles-write-variation.html:
```diff
@@ -0,0 +1,180 @@
+        }
+    });
+
+    WI.domManager.requestDocument((node) => {
```
I seem to recall we prefer getting the document inside the tests now. E.g. above line 33 above we can just `let documentNode = await WI.domManager.requestDocument()`. Then we don't need to have the `documentNode` shared variable, and future tests that could cause the document to change (unlikely here) would still get the correct document for the test.

patrickangle commented on LayoutTests/inspector/model/font-styles-write-variation.html:
```diff
@@ -0,0 +1,180 @@
+<div id="custom-axis-to-variation-settings-update"></div>
+
+</body>
+</html>
```
Style: Missing newline at end of file.

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -0,0 +1,340 @@
+        if (cssProperty)
+            return cssProperty;
+
+        // FIXME: Pick appropriate CSSStyleDeclaration from CSS cascade for font property.
```
Can we open a bug on bugzilla and note it here `// FIXME: <webkit.org/b/######> ...`

rcaliman-apple replied:
Filed and noted https://bugs.webkit.org/show_bug.cgi?id=250127

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/FontStyles.js:
```diff
@@ -0,0 +1,340 @@
+        this._authoredFontVariationSettingsMap = new Map;
+        this._effectiveWritablePropertyForNameMap = new Map;
+
+        // A significant change happens when there are different variation axes between style changes.
```
Would `different variation axis values` be a clearer way to say this?

rcaliman-apple replied:
I guess the comment didn't do its job :) 
But I'm concerned `different variation axis values` adds to the confusion because the next line says that changes to variation axis *values* isn't considered a significant change. Only different axes, identified by their tags, cause a significant change.

rcaliman-apple replied:
Reworded like so:
```
// A change in the number of axes or their tags is considered a significant change.
// A change to the value of a known axis is not considered a significant change.
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
```diff
@@ -188,6 +192,7 @@ body[dir=rtl] .details-section > .header::before {
     background-color: var(--background-color-intermediate);
 }
 
+
```
Oops?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+    grid-column: 2;
+    text-align: end;
+    color: var(--text-color-secondary);
+}
```
Style: Missing newline

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+.details-section > .content > .group > .row.font-variation {
```
Style: Missing copyright header

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+.details-section > .content > .group > .row.font-variation {
+    display: grid;
+    gap: 6px;
+    grid-template-columns: 36px 1fr minmax(max-content, 36px);
```
Seems like at the `36px` should be a variable with a name to make the columns easier to understand the purpose of.

Also we should take advantage of the fact we can name columns, which saves us from having a bunch of column numbers defined below as well. e.g. (but with better names 🥹)
```
grid-template-columns: [tag-name] 36px [slider] 1fr [value] minmax(max-content, 36px);
```
and then for the `.details-section > .content > .group > .row.font-variation > .value` rule you can just
```
grid-column: value;
```
and same for the other two columns as well. (The `grid-column: 1 / -1;` is probably fine though since it will always work no matter how many columns you add/rearrange)

rcaliman-apple replied:
Excellent advice! Will implement.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+    gap: 6px;
+    grid-template-columns: 36px 1fr minmax(max-content, 36px);
+    padding-block: 5px;
+    padding-inline: 20px;
```
Should these values match other types of rows (I suspect the answer is yes)? If so, why do we need to explicitly define them here, since the `.row` should already have them? Or if we are adding padding to existing padding, some variables and `calc` to make sure we add a consistent amount to the default padding so that if the other padding changes this follows suit?

rcaliman-apple replied:
A look at `DetailsSection.css` shows significant variance between padding for different types of rows.
The contents of `DetailsSectionRow` instances are very diverse: from read-only lists (Fonts), to tables (Node panel), to checkboxes and labels (Flex + Grid), etc. 

The variation axis control in Fonts warrants special padding to aid visual separation.

`padding-left: 12px` seems to be most prevalent among rows to somewhat align their contents with the arrow toggle icon of the section header. I'll use that value, but I'll keep a wide `padding-right` to avoid overlap with the scrollbar in its expanded state.

Here's 12px uniform inline padding where the axis values get partially covered:
<img width="321" alt="Screenshot 2023-01-05 at 12 02 23 PM" src="https://user-images.githubusercontent.com/105809798/210777524-c9d38f18-9e8c-4e5d-b86a-136c370577b1.png">

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+.details-section > .content > .group > .row.font-variation > .value {
+    grid-column: 3;
+    text-align: end;
+    font-variant-numeric: tabular-nums;
```
🤩

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+    grid-template-columns: 1fr 1fr;
+}
+
+.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"] {
```
Part of me wants to say this ranged slider control should be a separate JS class and CSS file, but this is probably fine for now.

rcaliman-apple replied:
I'll keep it in this file for now since it's a specialized input. 
When we need it for something else, we can extract it then.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,87 @@
+}
+
+.details-section > .content > .group > .row.font-variation > .variation-range > .variation-maxvalue {
+    grid-column: 2;
```
Would `grid-column: -1` make more sense here since this will always be the value at the end of the slider?

rcaliman-apple replied:
Using named columns instead per advice above.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/Variables.css:
```diff
@@ -219,6 +219,13 @@
     --image-button-navigation-item-width: 26px;
 
     --blackboxed-opacity: 0.5;
+
+    --slider-thumb-background: white;
+    --slider-thumb-background-active: hsl(0, 0%, 94%);
+    --slider-thumb-box-shadow: 0px 0 2px 0 hsl(0, 0%, 55%);
+    --slider-track-background: hsl(0, 0%, 88%);
+    --slider-track-box-shadow: inset 0 0 3px hsl(0, 0%, 70%);
+    --slider-track-tick-background: hsl(0, 0%, 70%);
```
Do we need to make any tweaks to the inactive style to match the inspiration in System Settings?

rcaliman-apple replied:
Is there an inactive state for sliders in the System Settings?

The current styles follow as closely as possible the styles in System Settings except the tone of the grays which have a brown tint in the system panels. For Web Inspector I stuck with simple grays.

patrickangle replied:
I guess there isn't 😅

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+          return 0.01;
+        } else if (delta <= 10) {
+          return 0.1;
+        }
```
Style: Single line `if`s don't need `{ }`. Also, because the first `if` is an early return, the `else` part of the second `if` is unnecessary.

Also, the `return`s should be intended four spaces.

rcaliman-apple replied:
Fixed.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+
+    _getAxisResolution(min, max)
+    {
+        let delta = parseInt(max, 10) - parseInt(min, 10);
```
The `min` and `max` from a `WI.FontVariationAxis` should always be a number, so it shouldn't be necessary to use `parseInt`. If it is necessary I'm very curious what causes it to be so.

rcaliman-apple replied:
They are numbers; not trying to cast from string to number here.

Since `min`/`max` can be float, math operation in JS can return weird results, ex: `0.1 + 0.2 = 0.30000000000000004`.

Here we only need to know the absolute delta between `min` and `max` to check in the conditions below.

`parseInt()` is not particularly _needed_ in this case. Just habit, I guess :)

There are some variable fonts with esoteric ranges, like [Skia](https://v-fonts.com/fonts/skia).

patrickangle replied:
Ah. If we just want an integer value we should probably use `Math.round()`. Also I'm not sure the floating point rounding error would cause much of an issue here since we are only using the range internally here to figure out how many fractional figures to display for a value.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+            this.element.appendChild(this._warningElement);
+        else
+            this._warningElement.remove();
+    }
```
Is there a way we can make this work without duplicating this logic from `DetailsSectionRow`?

rcaliman-apple replied:
I guess you meant `DetailsSectionSimpleRow` which also subclasses `DetailsSectionRow`.
I can move the setter and getter for `warningMessage` to `DetailsSectionRow` so both classes get it.

rcaliman-apple replied:
Done.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+
+        let {name, tag, value, defaultValue, minimumValue, maximumValue} = fontVariationAxis;
+
+        console.assert(typeof tag === "string" && tag.length === 4, "Invalid font variation axis tag", tag);
```
This assertion might make more sense in `FontVariationAxis.js` instead so that all clients of that class get this guarantee.

rcaliman-apple replied:
Good call.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+        this._inputRangeElement.setAttribute("min", minimumValue);
+        this._inputRangeElement.setAttribute("max", maximumValue);
+        this._inputRangeElement.setAttribute("value", value ?? defaultValue);
+        this._inputRangeElement.setAttribute("step", this._getAxisResolution(minimumValue, maximumValue));
```
All of these attributes have native JavaScript members we can use on the element. e.g.
```
this._inputRangeElement.type = "range";
```

rcaliman-apple replied:
Setting them as attributes helped me quickly debug at a glance by inspecting the node and observing values.
If you prefer it, I can set them as properties. Otherwise, I'll leave them as attributes for the same reason :)

rcaliman-apple replied:
Updated to properties on the object.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js:
```diff
@@ -0,0 +1,179 @@
+
+        this._labelElement = document.createElement("div");
+        this._labelElement.className = "label";
+        this.element.appendChild(this._labelElement);
```
In newer code we have favored appending the child at the same time we create it when possible. e.g.
```
this._labelElement = this.element.appendChild(document.createElement("div"));
this._labelElement.className = "label";
```

and ditto for the other elements create below as well

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -43,23 +50,42 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
-        this._fontPropertiesMap = fontMap.propertiesMap ?? new Map;
-        this._fontVariationsMap = fontMap.variationsMap ?? new Map;
-        this._fontFeaturesMap = fontMap.featuresMap ?? new Map;
+        // FIXME: Prevent refreshing FontStyles if we're the ones causing the changes.
```
Ditto about creating a bugzilla and referencing it here.

rcaliman-apple replied:
Filed and noted: https://bugs.webkit.org/show_bug.cgi?id=250128

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -35,6 +35,13 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        this._basicPropertyRowsMap = new Map;
+        this._basicPropertyNames = ["font-size", "font-style", "font-weight", "font-stretch"];
+
+        this._abortController = new AbortController;
```
I'm going to let you explain what an AbortController is in our meeting tomorrow 😅

patrickangle reviewed (commented):
I haven't finished reviewing yet (last file I need to finish is `FontDetailsPanel.js`), but looking good!

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -73,21 +99,77 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+
+        this._fontPropertiesMap = this._fontStyles.propertiesMap;
+        this._fontVariationsMap = this._fontStyles.variationsMap;
+        this._fontFeaturesMap = this._fontStyles.featuresMap;
```
Do we actually need to have local members here? Maybe we use some private getters instead if we don't want to type `_fontStyles` every time we edit?

rcaliman-apple replied:
Took your suggestion and went ahead making them actual private getters:

```
get #fontPropertiesMap()
{
    return this._fontStyles?.propertiesMap ?? new Map;   
}

get #fontVariationsMap()
{
    return this._fontStyles?.variationsMap ?? new Map;
}

get #fontFeaturesMap()
{
    return this._fontStyles?.featuresMap ?? new Map;
}
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -130,58 +208,55 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        let fontVariationAxis = fontProperty.variations?.values().next().value;
+        if (fontVariationAxis) {
+            // Ensure registered axes have a name; fallback to labels for their corresponding font properties.
+            fontVariationAxis.name ||= labelForTag[fontVariationAxis.tag] || "";
```
The name for the axis should be `null` from the backend if it isn't defined, and in general I think null is a better way to indicate "no value" than using an empty string, so I'd
```
fontVariationAxis.name ??= labelForTag[fontVariationAxis.tag];
```
And then handle the possibility that there may not be a name for an axis.

rcaliman-apple replied:
The empty string fallback was intentional to make this selector and CSS rule match:
```
.details-section > .content > .group > .row.font-variation > .label:not(:empty) {
    grid-column: 1 / -1; /* The variation axis name, if present, spans the full width of the grid container. */
}
```

~~But I guess I can explicitly handle the missing name in `FontVariationDetailsSectionRow.js` and not generate the markup for it.~~

Scratch that. I can use your suggestion and keep the selector as is. 

Assigning `this._labelElement.textContent = undefined` satisfies the selector. We get `undefined` from `labelForTag[fontVariationAxis.tag]` when there's no fallback axis name, for example for unnamed custom axes.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -130,58 +208,55 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
-        if (property.value === "normal")
-            return WI.UIString("Normal", "Normal @ Font Details Sidebar Property Value", "Property value for any `normal` CSS value.");
+        let fontProperty = this._fontPropertiesMap.get(propertyName);
+        let labelForTag = {
```
Nit: `const`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -130,58 +208,55 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
+        case "font-stretch":
+            return new WI.DetailsSectionSimpleRow(WI.UIString("Stretch", "Stretch @ Font Details Sidebar Property", "Property title for `font-stretch`."));
+            break;
+        }
```
Should we assert that we don't reach here?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css:
```diff
@@ -0,0 +1,118 @@
+}
+
+.details-section > .content > .group > .row.font-variation > .label:not(:empty) {
+    grid-column: 1 / -1; /* The variation axis name, if present, spans the full width of the grid container. */
```
This comment is probably redundant?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
```diff
@@ -130,58 +195,73 @@ WI.FontDetailsPanel = class FontDetailsPanel extends WI.StyleDetailsPanel
-    _formatSizeValue(property)
+    // Private
+
+    get #fontPropertiesMap()
```
When I said "private" I had initially meant "private" in the way we have typically handled class-private members (e.g. prefix them with `_`). I'm torn on this. On one hand, we should start using the actual private syntax, on the other hand we should probably resolve www.webkit.org/b/235859 before adopting private members, otherwise we begin to reduce the debug-ability of Web Inspector...

rcaliman-apple replied:
Agreed. Private member syntax is very nice to use, but perhaps it can wait until proper Web Inspector support.
It looked off that only some members were private and others were just `_` "private". I'll keep it uniform.

patrickangle reviewed (approved):
Awesome work. Just a few comments, but nothing show-stopping.

rcaliman-apple committed:
Web Inspector: Add editing capabilities for font variation axes in Fonts sidebar panel
https://bugs.webkit.org/show_bug.cgi?id=249535

Reviewed by Patrick Angle.

Variable fonts expose variation axes to control specific aspects of the font,
like thickness, stretch or slant. The Fonts sidebar panel already lists the
available variation axes for the primary font used on the selected node.

This patch augments the Fonts sidebar panel with editing controls to interactively
adjust the values for variation axes and get immediate visual feedback of the results.

There are two types of variation axes: registered and custom.

Registered axes values map to a corresponding CSS font property:
`wght` -> font-weight
`wdth` -> font-stretch
`ital` -> font-style
`slnt` -> font-style

Custom axes and their values are always written to `font-variation-settings`.

If a registered axis is declared in the value of `font-variation-settings`,
its value must always be written there instead of the corresponding font property
else it will be overriden by the value of `font-variation-settings`.

In this implementation, all edits made using the controls in the Fonts sidebar panel
are written to the selected node's inline style to ensure immediate visual feedback
at the expense of scoping the font style changes very narrowly to the selected node only.

A future iteration will improve this by identifying the appropriate CSS rules
in the cascade to write to.

* LayoutTests/inspector/model/font-calculate-properties-expected.txt:
* LayoutTests/inspector/model/font-calculate-properties.html:
* LayoutTests/inspector/model/font-styles-write-variation-expected.txt: Added.
* LayoutTests/inspector/model/font-styles-write-variation.html: Added.

Updated existing test to use the new `WI.FontStyles` model.
Added new test to validate variation axis writing logic.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Models/Font.js:
(WI.Font.prototype.variationAxis):
(WI.Font):
(WI.Font.prototype.calculateFontProperties): Deleted.
(WI.Font.prototype._calculateProperties): Deleted.
(WI.Font.prototype._calculateFontFeatureAxes): Deleted.
(WI.Font.prototype._calculateFontVariationAxes): Deleted.
(WI.Font.prototype._parseFontFeatureOrVariationSettings): Deleted.

Moved parsing-related logic from `WI.Font` to a new `WI.FontStyles` model.
This object maintains state of edits made to variation axes over time.
It handles where to write variation axis values.

* Source/WebInspectorUI/UserInterface/Models/FontStyles.js: Copied from Source/WebInspectorUI/UserInterface/Models/Font.js.
(WI.FontStyles):
(WI.FontStyles.prototype.get featuresMap):
(WI.FontStyles.prototype.get variationsMap):
(WI.FontStyles.prototype.get propertiesMap):
(WI.FontStyles.prototype.get significantChangeSinceLastRefresh):
(WI.FontStyles.fontPropertyForAxisTag):
(WI.FontStyles.axisValueToFontPropertyValue):
(WI.FontStyles.fontPropertyValueToAxisValue):
(WI.FontStyles.prototype.writeFontVariation):
(WI.FontStyles.prototype.refresh):
(WI.FontStyles.prototype._calculateFontProperties):
(WI.FontStyles.prototype._calculateProperties):
(WI.FontStyles.prototype._calculateFontFeatureAxes):
(WI.FontStyles.prototype._calculateFontVariationAxes):
(WI.FontStyles.prototype._parseFontFeatureOrVariationSettings):

Most contents of `WI.FontStyles` are copied over from `WI.Font` verbatim.

* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section > .content > .group:has(.row.font-variation)):
(.details-section > .content > .group > .row:is(.simple, .font-variation):has(.warning)):
(.details-section > .content > .group > .row:is(.simple, .font-variation) > .warning):
(.details-section > .content > .group > .row.simple:has(.warning)): Deleted.
(.details-section > .content > .group > .row.simple > .warning): Deleted.
* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel):
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.detached):
(WI.FontDetailsPanel.prototype.layout):
(WI.FontDetailsPanel.prototype.initialLayout):
(WI.FontDetailsPanel.prototype._createDetailsSectionRowForProperty):
(WI.FontDetailsPanel.prototype._formatPropertyValue):
(WI.FontDetailsPanel.prototype._handleFontVariationValueChanged):
(WI.FontDetailsPanel.prototype._formatSizeValue): Deleted.
(WI.FontDetailsPanel.prototype._formatStyleValue): Deleted.
(WI.FontDetailsPanel.prototype._formatSimpleSingleValue): Deleted.
(WI.FontDetailsPanel.prototype._formatVariationValue): Deleted.
(WI.FontDetailsPanel.prototype._createVariationValueElement): Deleted.
(WI.FontDetailsPanel.prototype._hasVariationValue): Deleted.

Inside `FontDetailsPanel.layout()` the structure for editing controls is created.
Inside `FontDetailsPanel.refresh()` the values for editing controls are updated.
This ensures the UI remains in sync when changes occur in the Styles sidebar panel
or new styles apply as a result of DOM mutations, matching media queries, etc.

* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.css: Added.
(.details-section > .content > .group > .row.font-variation):
(.details-section > .content > .group > .row.font-variation > .label:not(:empty)):
(.details-section > .content > .group > .row.font-variation > .tag):
(.details-section > .content > .group > .row.font-variation > .value):
(.details-section > .content > .group > .row.font-variation > .variation-range):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]::-webkit-slider-thumb):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]::-webkit-slider-thumb:active):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]::-webkit-slider-runnable-track):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]::before,):
(.details-section > .content > .group > .row.font-variation > .variation-range > input[type="range"]::after):
(.details-section > .content > .group > .row.font-variation > .variation-range > .variation-minvalue):
(.details-section > .content > .group > .row.font-variation > .variation-range > .variation-maxvalue):
* Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js: Added.
(WI.FontVariationDetailsSectionRow):
(WI.FontVariationDetailsSectionRow.prototype.get tag):
(WI.FontVariationDetailsSectionRow.prototype.set tag):
(WI.FontVariationDetailsSectionRow.prototype.get label):
(WI.FontVariationDetailsSectionRow.prototype.set label):
(WI.FontVariationDetailsSectionRow.prototype.get value):
(WI.FontVariationDetailsSectionRow.prototype.set value):
(WI.FontVariationDetailsSectionRow.prototype.get warningMessage):
(WI.FontVariationDetailsSectionRow.prototype.set warningMessage):
(WI.FontVariationDetailsSectionRow.prototype._formatAxisValueAsString):
(WI.FontVariationDetailsSectionRow.prototype._getAxisResolution):

`WI.FontVariationDetailsSectionRow` encapsulates the layout for editing controls:
the interactive slider, the track with min/max values, warnings for out-of-range values.

* Source/WebInspectorUI/UserInterface/Views/Variables.css:
(:root):
(@media (prefers-color-scheme: dark) :root):

Canonical link: https://commits.webkit.org/258503@main

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
```diff
@@ -180,6 +180,10 @@ body[dir=rtl] .details-section > .header::before {
     border-bottom: 1px solid var(--border-color);
 }
 
+.details-section > .content > .group:has(.row.font-variation) {
```
this kinda seems like a layering violation, no? why should `WI.DetailsSection` know about `WI.FontVariationDetailsSectionRow`?
